### ISSUE TITLE
Standardized meta data fields

### ISSUE THREAD

**Issue by almarklein** [imageio/imageio#382]
*See #362 for the meta-data overview issue.*

Note: Originally the shape of the meta dict was part of this issue, but I moved that to #502.

**The question: what fields to include as standard meta data (and how to name them)?**

Lists and names up for discussion!

I'd rather start small and extend later. Pretty sure:
* dtype (why?)
* description
* shape  (in pixels, so not necessarily equal to `array.shape`)
* spacing
* origin
* orientation / rotation
* Temporal spacing
* Acquisition date time
* axes ("xyz", "tzcyx", though don't imageio always return zyx?)
* nchannels
* Colorspace info (but we always return RGB, or doesnt it?

**Comment by almarklein** [comment]
An alternative might be to add fields in the root of the existing meta dict, and prefix them, like:
```py
im.meta['$shape']   # let's pray that no image format prefixes meta attributes with dollar signs
```

**Comment by almarklein** [comment]
cc @lschr @jni @jakirkham

**Comment by thewtex** [comment]
Hi @almarklein , this is great :+1: 

+1 for `origin` and a request for `direction` -- an orientation matrix is important, especially for medical images.

**Comment by almarklein** [comment]
What does `axes`, look like? A string being either `zyx` or `xyz`?

What does colorspace info look like, a tuple with descriptions like this?
```py
info["channels"] = ("red", "green", "blue")
```

> After I understood the third method, that became my favourite.

This would be that the main meta dict contains the standardized fields, plus one `raw` field for the original data. I like that approach the best as well, but unfortunately it would break backwards compatibility for imageio, which is why I'm leaning towards adding a new field to the existing metadata dict with a unique enough name.

**Comment by jni** [comment]
@almarklein

> what does axes look like?

A string which is a permutation of "tzyxc". For example, Leica microscopes save their files in acquisition order, which results in "tzcyx".

> what does colorspace info look like

Yeah, that's ok. We can leave this a bit more freeform. Multispectral imaging might have a tuple of wavelengths instead of colour names. In microscopy, people typically save the emission wavelength.

> unfortunately it would break backwards compatibility for imageio

GAH. Well, this is always a hard choice, but given my thus-far light dependency on imageio, I'd still favour ripping the bandaid hard. =P You could deprecate it quite effectively by turning `.meta` into a property instead of a dictionary on the Image object.

**Comment by almarklein** [comment]
> by turning .meta into a property 

What do you mean? `.meta` is now a (dict) property of the Image object.

**Comment by thewtex** [comment]
The image `direction` is defined to be the direction cosine matrix that defines an image's orientation about its origin. Or, from the [ITK Software Guide](https://itk.org/ITKSoftwareGuide/html/Book1/ITKSoftwareGuide-Book1ch4.html#x45-500004.1):

> The image direction matrix represents the orientation relationships between the image samples and physical space coordinate systems. The image direction matrix is an orthonormal matrix that describes the possible permutation of image index values and the rotational aspects that are needed to properly reconcile image index organization with physical space axis. The image directions is a NxN matrix where N is the dimension of the image. An identity image direction indicates that increasing values of the 1st, 2nd, 3rd index element corresponds to increasing values of the 1st, 2nd and 3rd physical space axis respectively, and that the voxel samples are perfectly aligned with the physical space axis.

> For a given index I3X1, the physical location P3X1 is calculated as follows:
￼
![image](https://user-images.githubusercontent.com/25432/46479359-e08e4780-c7bc-11e8-8cf4-9f383c39db34.png)

> where D is an orthonormal direction cosines matrix and S is the image spacing diagonal matrix.

In medical imaging, it is common to roughly align the data collection axes with the patient, i.e. superior-inferior, anterior-posterior, left-right. The direction matrix is also often used to flip or rotate the image content appropriately.

The direction, like spacing and origin, are critical for operations like registration.

Since the pixel collection axes are not necessarily aligned with the world coordinate axes, pixel axes labeling with `ijk` as opposed to `xyz` can help avoid confusion. I know `xyz` is conventional in microscopy and ImageJ, but the pixel versus world coordinate system distinction is quite helpful to avoid confusion.

**Comment by almarklein** [comment]
@jni, maybe we can do something like this:
```py
@property
def meta(self):
    """A dict containing the raw meta-data, as it always was."""
    ...

@property
def info(self):
    """A dict containing standardized meta-data, including a field `raw` which corresponds to self.meta."""
    ...
```
This way there should be no deprecation issues at all, *and* the info dict would match skimage's approach (if that's what skimage will adopt).

**Comment by jni** [comment]
Works for me! =)

**Comment by GenevieveBuckley** [comment]
Image metadata fields I use most day-to-day include:
* frame interval for time series data (length between one timepoint and the next)
* voxel dimensions in x, y and z (it's unclear to me if this is the same thing as "spacing")
* shape of image in pixels (already on the list above)

(Note that some types of transformations can change the effective voxel size. I am not sure how best to harndle this type of situation in terms of metadata.)

**Comment by jni** [comment]
@GenevieveBuckley 

> it's unclear to me if this is the same thing as "spacing"

I used to favour "voxel size" as the terminology until I read [this paper](http://alvyray.com/Memos/CG/Microsoft/6_pixel.pdf). "spacing" makes more sense with the "point sample" model of pixels, and indeed it generalises to frame intervals (though perhaps those are best kept separate anyway).

**Comment by almarklein** [comment]
@jni I'm starting to get the impression that the whole main about this meta data is to be able to position and orient the image data in 3D space (or space-time).

The one exception is colorspace (I opened #504 to discuss that).

**Comment by FirefoxMetzger** [comment]
closed via #739
